### PR TITLE
added reduced chance of collision

### PR DESCRIPTION
**Pull request by TheForestSystem**

### CODE REVIEW

File: src/GeneratorUtils.php [hunk 1]
```diff
@@ -4,19 +4,25 @@
...
-            throw new InvalidArgumentException("The character array must have two items!");
-        }
+        $charactersLength = count($characters) - 1;
+        $string = '';
 
-        for ($i = 0; $i < $length; $i++) {
-            $id .= $characters[rand(0, count($characters)-1)];
-        }
+        for ($i = 0; $i < $length; $i++) 
+            $string .= $characters[random_int(0, $charactersLength)];
```
Comment (CommandString) [inline src/GeneratorUtils.php]: This breaks PSR

File: src/GeneratorUtils.php [hunk 2]
```diff
@@ -1,22 +1,32 @@
...
-    public static function uuid(int $length = 16, ?array $characters = null): string
+    public static function uuid(int $length = 16): string
     {
-        $characters = $characters ?? array_merge(range("A", "Z"), range("a", "z"), range(0, 9));
-        $id = "";
+        $salt = random_bytes($length);
+        $time = microtime(true);
+        $hash = hash('sha256', $salt . $time);
+        return $hash;
+    }
```
Comment (CommandString) [inline src/GeneratorUtils.php:12]: Last I checked hashes return the same length so the parameter is misleading. [How long is the SHA256 hash?](https://stackoverflow.com/questions/2240973/how-long-is-the-sha256-hash/2241014#2241014)

This also breaks backwards compatibility which I want to avoid.

File: src/GeneratorUtils.php [hunk 3]
```diff
@@ -1,22 +1,32 @@
...
+        $charactersLength = count($characters) - 1;
+        $string = '';
+
         for ($i = 0; $i < $length; $i++) {
-            $id .= $characters[rand(0, count($characters)-1)];
+            $string .= $characters[random_int(0, $charactersLength)];
         }
 
-        return $id;
+        return $string;
```
Comment (CommandString) [inline src/GeneratorUtils.php:30]: What not foreach the $characters array again? Should be using [mt_rand](https://www.php.net/manual/en/function.mt-rand.php). Also, why was the exception removed?

~~I understand this probably was my original code.~~

**Review by CommandString — CHANGES_REQUESTED** [review]

File: src/GeneratorUtils.php (hunk 2 above)
Comment (TheForestSystem) [reply src/GeneratorUtils.php:12]: > This also breaks backwards compatibility which I want to avoid.

how so?

File: src/GeneratorUtils.php (hunk 3 above)
Comment (TheForestSystem) [reply src/GeneratorUtils.php:30]: > What not foreach

can you rephrase? I'm confused.

File: src/GeneratorUtils.php (hunk 2 above)
Comment (CommandString) [reply src/GeneratorUtils.php:12]: Method performs differently now

File: src/GeneratorUtils.php (hunk 3 above)
Comment (CommandString) [reply src/GeneratorUtils.php:30]: Oh wait I'm sorry I had a moment. Ignore the foreach suggestion

File: src/GeneratorUtils.php (hunk 2 above)
Comment (TheForestSystem) [reply src/GeneratorUtils.php:12]: > Method performs differently now

it does the exact same thing but in a different way. 
`2*3` vs `3*2`

Comment (CommandString) [reply src/GeneratorUtils.php:12]: Can't supply custom characters now. Also, you never addressed the fact this will always return 256 bits (64 characters).

Comment (TheForestSystem) [reply src/GeneratorUtils.php:12]: > you never addressed the fact this will always return 256 bits (64 characters).

`random_bytes($length);`
yes I did?

Comment (CommandString) [reply src/GeneratorUtils.php:12]: You're returning a hash. That hash will **ALWAY** return 256 bits (64 characters) as I mentioned previously. The input does **NOT** affect the output like it does in encryption.
